File: Python/Objetos/Terminal/Movil.py
'''
Created on 7 feb. 2019

@author: d18cralr
'''

class Terminal:
    
    '''
    Constructor
    '''
    def __init__(self,numero):
        self.__numeroTelefono=numero
        self.__tiempoLlamada=0
    
    @property #Se puede usar como un "getter"
    def numeroTelefono(self):
        return self.__numeroTelefono
    
    @numeroTelefono.setter #setter de lado
    def numeroTelefono(self):
        self.__numeroTelefono
        
    @property #Se puede usar como un "getter"
    def tiempoLlamada(self):
        return self.__tiempoLlamada
    
    @tiempoLlamada.setter #setter de lado
    def tiempoLlamada(self,tiempo):
        self.__tiempoLlamada=tiempo
    
    '''
    Método llamar
    @param t numero de teléfono
    @param tiempo tiempo de la conversación
    '''    
    def llamar(self,t,tiempo):
        self.__tiempoLlamada += tiempo #Acumular el tiempo total que ha hablado con un terminal
        t.tiempoLlamada+=tiempo #Acumular el tiempo que se le pasa a un terminal como parametro
    
    '''
    Método toString 
    '''
    def __str__(self):
        return f"nº:{self.__numeroTelefono}-{self.__tiempoLlamada} s."
 
class Movil(Terminal):
    
    tarifa=""
    totalTarifa=0
 
    def __init__(self,numero,tarifa):
        super().__init__(numero)
        self.__tarifa=tarifa
        self.__totalTarifa=0
    
    def llamar(self, t, tiempo):
        Terminal.llamar(self, t, tiempo)
        minutos=tiempo/60
        
        if (self.__tarifa=="rata"):
            self.__totalTarifa += minutos*0.06
        
        elif self.__tarifa=="mono":
            self.__totalTarifa+=minutos*0.12
            
        elif self.__tarifa=="bisonte":
            self.__totalTarifa+=minutos*0.30
            
        
    def __str__(self):
        return str(super().__str__())+f"-tarificados {self.__totalTarifa} euros."
        
        
      
if __name__=="__main__":
    
    m1=Movil("111 111 111","rata")
    m2=Movil("222 222 222","mono")
    m3=Movil("333 333 333","bisonte")
    
    print(m1)
    print(m2)
    print("Tiempo tras llamada: ")
    m1.llamar(m2,320)
    m1.llamar(m3,200)
    m2.llamar(m3, 550)
    print(m1)
    print(m2)
    print(m3)
   
    

File: Java/Tema2-POO/Terminal/Movil.java
package Terminal;

import java.text.DecimalFormat;

/**
 * Implementa la clase Movil como subclase de Terminal (la clase del ejercicio
 * anterior que ya no hace falta modificar). Cada móvil lleva asociada una
 * tarifa que puede ser "rata", "mono" o "bisonte". El coste por minuto es de 6,
 * 12 y 30 céntimos respectivamente. Se tarifican los segundos exactos.
 * Obviamente, cuando un móvil llama a otro, se le cobra al que llama, no al que
 * recibe la llamada. A continuación se proporciona el contenido del main y el
 * resultado que debe aparecer por pantalla. Para que el total tarificado
 * aparezca con dos decimales, puedes utilizar DecimalFormat. Contenido del main
 *
 *
 * @author Rafael Miguel Cruz Álvarez
 * @version 1.0
 */


public class Movil extends Terminal{

  private String tarifa;
  private double totalTarifa;
  
  public Movil(String numero,String tarifa) {
    super(numero);
    this.tarifa = tarifa;
    this.totalTarifa = 0;
  }
  
  @Override
  public void llamar(Terminal t, int tiempo) {
    super.llamar(t, tiempo);
    double minutos = (double) tiempo/60;
    
    switch (this.tarifa) {
      case "rata":
        this.totalTarifa += minutos*0.06;
        break;
      case "mono":
        this.totalTarifa+=minutos*0.12;
        break;
      case "bisonte":
        this.totalTarifa+=minutos*0.30;
        break;
    }
  }

  /* (non-Javadoc)
   * @see java.lang.Object#toString()
   */
  @Override
  public String toString() {
    DecimalFormat formateador = new DecimalFormat("0.00");
    return super.toString()+" - tarificados "+formateador.format(this.totalTarifa)+" euros";
  }
  
  
  
}
